battery.py

```diff
@@ -0,0 +1,67 @@
+import os
+import power
+
+battery_color_good = [  0,230,0]
+battery_color_ok   = [255,215,0]
+battery_color_bad  = [255,  0,0]
+
+BATTERY_FLASH = [None,None,None,[3],[2,3],[1,2],[2,3],[1,2],[1],None,None,None,None]
+
+def get_bat_color():
+    """
+    Function determines the color of the battery indicator. Colors can be set in config.
+    Voltage threshold's are currently estimates as voltage isn't that great of an indicator for
+    battery charge.
+    :return: false if old firmware, RGB color array otherwise
+    """
+    try:
+        v = os.read_battery()
+        if v > 3.8:
+            return battery_color_good
+        if v > 3.6:
+            return battery_color_ok
+        return battery_color_bad
+    except AttributeError:
+        return False
+
+
+def render_battery(disp):
+    """
+    Adds the battery indicator to the display. Does not call update or clear so it can be used in addition to
+    other display code.
+    :param disp: open display
+    """
+    c = get_bat_color()
+    if not c:
+        return
+    disp.line(141, 62, 155, 62, col = c)
+    disp.line(155, 62, 155, 69, col = c) # line bug
+    disp.line(155, 68, 141, 68, col = c)
+    disp.line(141, 68, 141, 62, col = c)
+    disp.rect(156, 64, 158, 67, filled = True, col = c)
+    try:
+        charging = (power.read_chargein_voltage() > 1)
+    except AttributeError:
+        charging = False
+    try:
+        v = os.read_battery()
+        fillWidth = round(max(min(v - 3.6, 0.44), 0) * 32.5)
```

Comment:
those magic floats look a little interesting. I'll have to look into this a little more.

If you have a specific reasoning behind this It'd be great to hear that if you have some time.
Else, I'll just see if I can figure this out myself.

Reply:
That's because of the battery voltage levels. Also the general idea taken from the digital clock (see line 4), where empty is defined as 3.6 volts, and full around 4.1-ish volts. With that in mind, it is quite simple to do rule of three to get over to width values, as the "full width" is also known to be 14. So the factors range between the minimal voltage (3.6) - interpreted as zero width - and the maximal voltage (4.04) - interpreted as width 14. Finally the searched counterpart factor is 32.5 to achieve this with only one multiplication and no complex calculations